operator/rpcoperator.go

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package operator
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/perun-network/erdstall/tee"
+)
+
+type (
+	// WireAPI contains all functions that a client is interested in.
+	// It is called `WireAPI` since the client will interact with an
+	// implementation of it over the wire.
+	WireAPI interface {
+		Send(tee.Transaction) error
+		SubscribeProofs(common.Address) (ProofSub, error)
+	}
+
+	// RPCOperator will be exposed to the client as a websocket RPC Server.
+	// It implements `WireAPI`.
+	RPCOperator struct {
+		mtx     sync.Mutex // protects all
+		enclave tee.Enclave
+
+		subs map[common.Address]*ProofSub
+	}
+
+	// ProofSub is a subscription on TEE proofs.
+	ProofSub struct {
+		deposits chan tee.DepositProof
+		balances chan tee.BalanceProof
+		quit     chan struct{}
+	}
+)
+
+var _ WireAPI = (*RPCOperator)(nil)
+
+func NewRPCOperator(enclave tee.Enclave) *RPCOperator {
+	return &RPCOperator{
+		enclave: enclave,
+		subs:    make(map[common.Address]*ProofSub),
+	}
+}
+
+func (o *RPCOperator) Send(tx tee.Transaction) error {
+	o.mtx.Lock()
+	defer o.mtx.Unlock()
+	return o.enclave.ProcessTXs(&tx)
+}
+
+// SubscribeProofs returns a subscription on TEE proofs for the given address.
+// The subscription buffers the most recent proof until the client retrieves it.
+func (o *RPCOperator) SubscribeProofs(addr common.Address) (ProofSub, error) {
+	o.mtx.Lock()
+	defer o.mtx.Unlock()
+	log.WithField("who", addr.Hex()).Debug("Subscribed to proofs")
+	sub, ok := o.subs[addr]
+	if !ok {
+		o.subscribe(addr)
+		sub = o.subs[addr]
+	}
+	return *sub, nil
+}
+
+// subscribe is an internal implementation detail and should not be called.
+func (o *RPCOperator) subscribe(addr common.Address) ProofSub {
+	o.subs[addr] = newProofSub()
+	return *o.subs[addr]
+}
+
+func (o *RPCOperator) PushDepositProof(proof tee.DepositProof) {
```

Comment:
Do we really need to export the `Push...` methods? I mean, a subscriber outside this package isn't pushing new proofs into the sub, right? Same for balance proofs.

Reply:
It is used in a Black Box test and must therefore sadly be exported.